src/Core/BufferLine.re

```diff
@@ -33,24 +37,130 @@ type t = {
+  let skiaTypefaceEqual = (tf1, tf2) =>
+    Skia.Typeface.getUniqueID(tf1) == Skia.Typeface.getUniqueID(tf2);
+
+  // We want to cache measurements tied to typefaces and characters, since text measuring is expensive
+  module SkiaTypefaceUcharHashable = {
+    // This is of the form typeface, size, uchar
+    type t = (Skia.Typeface.t, float, Uchar.t);
+
+    let equal = ((tf1, s1, uc1), (tf2, s2, uc2)) =>
+      skiaTypefaceEqual(tf1, tf2) && s1 == s2 && Uchar.equal(uc1, uc2);
+
+    let hash = ((tf, s, uc)) =>
+      Int32.to_int(Skia.Typeface.getUniqueID(tf))
+      + Uchar.hash(uc)
+      + Hashtbl.hash(s);
+  };
+
+  module MeasureResult = {
+    type t = float;
+
+    let weight = _ => 1;
+  };
+
+  module MeasurementsCache =
+    Lru.M.Make(SkiaTypefaceUcharHashable, MeasureResult);
+
+  let measurementsCache =
+    MeasurementsCache.create(~initialSize=1024, 128 * 1024);
+
+  // Create a paint to measure the character with
+  let paint = Skia.Paint.make();
+  Skia.Paint.setTextEncoding(paint, GlyphId);
+  Skia.Paint.setLcdRenderText(paint, true);
+  Skia.Paint.setAntiAlias(paint, true);
+
+  let getCharacterWidth = (~indentation: IndentationSettings.t, uchar) =>
+    if (Uchar.equal(uchar, tab)) {
+      indentation.tabSize;
     } else {
-      Uucp.Break.tty_width_hint(c);
+      // Some unicode codepoints are double width, like many CJK characters
+      Uucp.Break.tty_width_hint(
+        uchar,
+      );
+    };
+
+  // The measure result is of the form (characterWidth, pixelWidth)
+  let measure = (~typeface: Skia.Typeface.t, ~cache: t, substr, uchar) =>
+    switch (
+      MeasurementsCache.find(
+        (typeface, cache.font.fontSize, uchar),
+        measurementsCache,
+      )
+    ) {
+    | Some(pixelWidth) =>
+      MeasurementsCache.promote(
+        (typeface, cache.font.fontSize, uchar),
+        measurementsCache,
+      );
+      Log.debugf(m =>
+        m(
+          "MeasurementCache : Hit! Typeface : %s, Uchar: %s (%d)",
+          Skia.Typeface.getFamilyName(typeface),
+          Zed_utf8.singleton(uchar),
+          Uchar.to_int(uchar),
+        )
+      );
+      let characterWidth =
+        getCharacterWidth(~indentation=cache.indentation, uchar);
+      (characterWidth, pixelWidth);
+    | None =>
+      Log.debugf(m =>
+        m(
+          "MeasurementCache : Miss! Typeface : %s, Uchar: %s (%d)",
+          Skia.Typeface.getFamilyName(typeface),
+          Zed_utf8.singleton(uchar),
+          Uchar.to_int(uchar),
+        )
+      );
+      // The width in monospaced characters
+      let characterWidth =
+        getCharacterWidth(~indentation=cache.indentation, uchar);
+      Skia.Paint.setTypeface(paint, typeface);
+
+      // When the character is a tab, we have to make sure
+      // we offset the correct amount.
+      let pixelWidth =
+        if (Uchar.equal(uchar, tab)) {
+          Skia.Paint.setTextEncoding(paint, Utf8);
+          let spaceSize = Skia.Paint.measureText(paint, "  ", None);
```

Comment:
Maybe it makes sense to just cache this as part of the `Font.t` object? Seems like we'd measure this a lot!

Reply:
Great idea. I just renamed `Font.t`'s `measuredWidth` to `spaceWidth` and set it to the width of a space character. Since fonts aren't always monospaced anymore, I think `measuredWidth` is slightly misleading.